pkg/storage/stores/shipper/gateway_client.go

```diff
@@ -136,6 +211,48 @@ func (s *GatewayClient) doQueries(ctx context.Context, queries []chunk.IndexQuer
 	return nil
 }
 
+// ringModeDoQueries prepares an index query to be sent to the Index Gateway, and then sends it
+// using the clientDoQueries implementation.
+//
+// The preparation and sending phase includes:
+// 1. Extracting the tenant name from the query.
+// 2. Fetching different Index Gateway instances assigned to the extracted tenant.
+// 3. Iterating in parallel over all fetched Index Gateway instances, getting their gRPC connections
+//  from the pool and invoking clientDoQueries using their client.
+func (s *GatewayClient) ringModeDoQueries(ctx context.Context, gatewayQueries []*indexgatewaypb.IndexQuery, queryKeyQueryMap map[string]chunk.IndexQuery, callback chunk.QueryPagesCallback) error {
+	userID, err := tenant.TenantID(ctx)
+	if err != nil {
+		return errors.Wrap(err, "index gateway client get tenant ID")
+	}
+
+	bufDescs, bufHosts, bufZones := ring.MakeBuffersForGet()
+
+	key := util.TokenFor(userID, "" /* labels */)
+	rs, err := s.ring.Get(key, ring.WriteNoExtend, bufDescs, bufHosts, bufZones)
+	if err != nil {
+		return errors.Wrap(err, "index gateway get ring")
+	}
+
+	addrs := rs.GetAddresses()
+	for _, addr := range addrs {
```

Comment:
Let us add some randomization to try different addresses randomly, otherwise we will go to same gateway each time for a user.

Reply:
Great catch. I pushed one commit addressing this, WDYT?